src/graphcast_agent/message_typing.rs

```diff
@@ -209,31 +216,76 @@ impl<
         &self,
         registry_subgraph: &str,
         network_subgraph: &str,
-        local_graphcast_id: String,
+        local_sender_id: String,
+        id_validation: IdentityValidation,
     ) -> Result<&Self, BuildMessageError> {
-        let graphcast_id = self.recover_sender_address()?;
-        if graphcast_id == local_graphcast_id {
-            Err(BuildMessageError::InvalidFields(anyhow!(
-                "Message is from self, drop message"
-            )))
-        } else {
-            let indexer_address =
-                query_registry_indexer(registry_subgraph.to_string(), graphcast_id)
+        match id_validation {
+            IdentityValidation::NoCheck => (),
+            IdentityValidation::ValidAddress => {
+                let _ = self.remote_account(local_sender_id)?;
+            }
+            IdentityValidation::GraphcastRegistered => {
+                let claimed_account = self.remote_account(local_sender_id)?;
+                // Simply check if the message signer is registered at Graphcast Registry, make no validation on Graph Account field
+                let _ = claimed_account
+                    .account_from_registry(registry_subgraph)
                     .await
+                    .map(|verified_account| {
+                        if verified_account.account() == claimed_account.account() {Ok(verified_account)} else {
```

Comment:
I think those are cloning the `account: String` right?

I see that you have multiple getter/accessor methods that clone to not have to deal with lifetimes, but I think I would prefer to clone only when it's needed, so it's more clear to whoever is accessing the data.